<a href="https://colab.research.google.com/github/r3tr0devilz/Data-Analysis-Projects/blob/main/Customer_care_churn_case_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Life time value Analysis. Part 2

Life time value is the total worth to a business of a customer over the while period of their relationship.

In [ ]:
import sqlite3
import pandas as pd

In [ ]:
conn = sqlite3.connect("churn (1).db")
cur = conn.cursor()

In [ ]:
df = pd.read_sql("SELECT * FROM churn_all", conn)

In [ ]:
df.head()

CustomerID  Gender SeniorCitizen Partner Dependents       State   Latitude  \
0  3668-QPYBK    Male            No      No         No  California  33.964131   
1  9237-HQITU  Female            No      No        Yes  California  34.059281   
2  9305-CDSKC  Female            No      No        Yes  California  34.048013   
3  7892-POOKP  Female            No     Yes        Yes  California  34.062125   
4  0280-XJGEX    Male            No      No        Yes  California  34.039224   

    Longitude  ZipCode PhoneService  ... TechSupport StreamingTV  \
0 -118.272783    90003          Yes  ...          No          No   
1 -118.307420    90005          Yes  ...          No          No   
2 -118.293953    90006          Yes  ...          No         Yes   
3 -118.315709    90010          Yes  ...         Yes         Yes   
4 -118.266293    90015          Yes  ...          No         Yes   

  StreamingMovies Tenure        Contract              PaymentMethod  \
0              No      2  Month-to-month               Mailed check   
1              No      2  Month-to-month           Electronic check   
2             Yes      8  Month-to-month           Electronic check   
3             Yes     28  Month-to-month           Electronic check   
4             Yes     49  Month-to-month  Bank transfer (automatic)   

  PaperlessBilling MonthlyCharges  TotalCharges Churn  
0              Yes          53.85        108.15   Yes  
1              Yes          70.70        151.65   Yes  
2              Yes          99.65         820.5   Yes  
3              Yes         104.80       3046.05   Yes  
4              Yes         103.70        5036.3   Yes  

[5 rows x 25 columns]

##For those customers who unsubscribed the service, what was their average LTV? and how long did they usually stay in the service?


In [ ]:
## Extract those who unsubscribed the serviced (Churn = Yes)

churn_df = pd.read_sql("SELECT * FROM churn_all where Churn='Yes'",conn)

In [ ]:
## Examine the distribution of TotalCharges
churn_df["TotalCharges"].describe()

count    1869.000000
mean     1531.796094
std      1890.822994
min        18.850000
25%       134.500000
50%       703.550000
75%      2331.300000
max      8684.800000
Name: TotalCharges, dtype: float64

20% of the charges are very high

In [ ]:
## Find 80th percentile of the data in TotalCharge
churn_df.TotalCharges.quantile(0.8)

2840.4100000000003

In [ ]:
pd.read_sql("select * from churn_all where TotalCharges <= 2840.41", conn)

CustomerID  Gender SeniorCitizen Partner Dependents       State  \
0     3668-QPYBK    Male            No      No         No  California   
1     9237-HQITU  Female            No      No        Yes  California   
2     9305-CDSKC  Female            No      No        Yes  California   
3     4190-MFLUW  Female            No     Yes         No  California   
4     8779-QRDMV    Male           Yes      No         No  California   
...          ...     ...           ...     ...        ...         ...   
4734  8456-QDAVC    Male            No      No         No  California   
4735  7750-EYXWZ  Female            No      No         No  California   
4736  2569-WGERO  Female            No      No         No  California   
4737  6840-RESVB    Male            No     Yes        Yes  California   
4738  4801-JZAZL  Female            No     Yes        Yes  California   

       Latitude   Longitude  ZipCode PhoneService  ...          TechSupport  \
0     33.964131 -118.272783    90003          Yes  ...                   No   
1     34.059281 -118.307420    90005          Yes  ...                   No   
2     34.048013 -118.293953    90006          Yes  ...                   No   
3     34.066367 -118.309868    90020          Yes  ...                  Yes   
4     34.023810 -118.156582    90022           No  ...                   No   
...         ...         ...      ...          ...  ...                  ...   
4734  32.852947 -114.850784    92283          Yes  ...                   No   
4735  34.159534 -116.425984    92284           No  ...                  Yes   
4736  34.341737 -116.539416    92285          Yes  ...  No internet service   
4737  34.667815 -117.536183    92301          Yes  ...                  Yes   
4738  34.167800 -116.864330    92305           No  ...                   No   

              StreamingTV      StreamingMovies Tenure        Contract  \
0                      No                   No      2  Month-to-month   
1                      No                   No      2  Month-to-month   
2                     Yes                  Yes      8  Month-to-month   
3                      No                   No     10  Month-to-month   
4                      No                  Yes      1  Month-to-month   
...                   ...                  ...    ...             ...   
4734                  Yes                   No     19  Month-to-month   
4735                  Yes                  Yes     12        One year   
4736  No internet service  No internet service     72        Two year   
4737                  Yes                  Yes     24        One year   
4738                   No                   No     11  Month-to-month   

                  PaymentMethod PaperlessBilling MonthlyCharges  TotalCharges  \
0                  Mailed check              Yes          53.85        108.15   
1              Electronic check              Yes          70.70        151.65   
2              Electronic check              Yes          99.65        820.50   
3       Credit card (automatic)               No          55.20        528.35   
4              Electronic check              Yes          39.65         39.65   
...                         ...              ...            ...           ...   
4734  Bank transfer (automatic)              Yes          78.70       1495.10   
4735           Electronic check               No          60.65        743.30   
4736  Bank transfer (automatic)              Yes          21.15       1419.40   
4737               Mailed check              Yes          84.80       1990.50   
4738           Electronic check              Yes          29.60        346.45   

     Churn  
0      Yes  
1      Yes  
2      Yes  
3      Yes  
4      Yes  
...    ...  
4734    No  
4735    No  
4736    No  
4737    No  
4738    No  

[4739 rows x 25 columns]

In [ ]:
churn_df.query("TotalCharges <= 2840.41")

CustomerID  Gender SeniorCitizen Partner Dependents       State  \
0     3668-QPYBK    Male            No      No         No  California   
1     9237-HQITU  Female            No      No        Yes  California   
2     9305-CDSKC  Female            No      No        Yes  California   
5     4190-MFLUW  Female            No     Yes         No  California   
6     8779-QRDMV    Male           Yes      No         No  California   
...          ...     ...           ...     ...        ...         ...   
1863  1122-JWTJW    Male            No     Yes        Yes  California   
1864  1699-HPSBG    Male            No      No         No  California   
1865  8775-CEBBJ  Female            No      No         No  California   
1866  6894-LFHLY    Male           Yes      No         No  California   
1868  8361-LTMKD    Male           Yes     Yes         No  California   

       Latitude   Longitude  ZipCode PhoneService  ... TechSupport  \
0     33.964131 -118.272783    90003          Yes  ...          No   
1     34.059281 -118.307420    90005          Yes  ...          No   
2     34.048013 -118.293953    90006          Yes  ...          No   
5     34.066367 -118.309868    90020          Yes  ...         Yes   
6     34.023810 -118.156582    90022           No  ...          No   
...         ...         ...      ...          ...  ...         ...   
1863  32.698964 -115.886656    92259          Yes  ...          No   
1864  33.745746 -116.514215    92264          Yes  ...         Yes   
1865  32.790282 -115.689559    92273          Yes  ...          No   
1866  34.264124 -114.717964    92280          Yes  ...          No   
1868  34.550820 -117.122396    92307          Yes  ...          No   

     StreamingTV StreamingMovies Tenure        Contract  \
0             No              No      2  Month-to-month   
1             No              No      2  Month-to-month   
2            Yes             Yes      8  Month-to-month   
5             No              No     10  Month-to-month   
6             No             Yes      1  Month-to-month   
...          ...             ...    ...             ...   
1863          No              No      1  Month-to-month   
1864         Yes              No     12        One year   
1865          No              No      9  Month-to-month   
1866          No              No      1  Month-to-month   
1868          No              No      4  Month-to-month   

                  PaymentMethod PaperlessBilling MonthlyCharges  TotalCharges  \
0                  Mailed check              Yes          53.85        108.15   
1              Electronic check              Yes          70.70        151.65   
2              Electronic check              Yes          99.65        820.50   
5       Credit card (automatic)               No          55.20        528.35   
6              Electronic check              Yes          39.65         39.65   
...                         ...              ...            ...           ...   
1863               Mailed check              Yes          70.65         70.65   
1864           Electronic check              Yes          59.80        727.80   
1865  Bank transfer (automatic)              Yes          44.20        403.35   
1866           Electronic check              Yes          75.75         75.75   
1868               Mailed check              Yes          74.40        306.60   

     Churn  
0      Yes  
1      Yes  
2      Yes  
5      Yes  
6      Yes  
...    ...  
1863   Yes  
1864   Yes  
1865   Yes  
1866   Yes  
1868   Yes  

[1495 rows x 25 columns]

In [ ]:
# Divide the data by the 80th percentile of the data from the TotalCharges variable
total_charges_under80 = churn_df.query("TotalCharges <= 2840.41")
total_charges_above80 = churn_df.query("TotalCharges > 2840.41")

In [ ]:
# Show the distribution of people under the 80th percentile
total_charges_under80.TotalCharges.describe()

count    1495.000000
mean      711.265819
std       766.848197
min        18.850000
25%        85.025000
50%       371.650000
75%      1128.225000
max      2839.650000
Name: TotalCharges, dtype: float64

In [ ]:
# Show the distribution of people above the 80th percentile
total_charges_above80.TotalCharges.describe()

count     374.000000
mean     4811.723262
std      1436.724288
min      2841.550000
25%      3522.462500
50%      4571.100000
75%      5891.212500
max      8684.800000
Name: TotalCharges, dtype: float64

In [ ]:
# Show the distribution of the tenure of people above the 80th percentile
total_charges_above80.Tenure.describe()

count    374.000000
mean      50.133690
std       12.334841
min       27.000000
25%       40.000000
50%       49.500000
75%       60.000000
max       72.000000
Name: Tenure, dtype: float64

In [ ]:
# Show the distribution of the tenure of people under the 80th percentile
total_charges_under80.Tenure.describe()

count    1495.000000
mean        9.935117
std        10.742349
min         1.000000
25%         1.000000
50%         6.000000
75%        15.000000
max        61.000000
Name: Tenure, dtype: float64

The average LTV of the 80% of unsubscribers is 711 dollars, and their duration is close to 10 months. On the other hand, the average LTV of the 20% of     unsubscribers is 4811 dollars, and their duration is close to 50 months.

In [ ]:
# under 80
pd.read_sql("select count(1) from churn_all where TotalCharges <= 2840.41 and Churn = 'Yes' group by PhoneService", conn)

count(1)
0       161
1      1334

In [ ]:
df["Gender"].value_counts() # example

Male      3555
Female    3488
Name: Gender, dtype: int64

In [ ]:
# under 80
total_charges_under80.PhoneService.value_counts()

Yes    1334
No      161
Name: PhoneService, dtype: int64

In [ ]:
# Example of shape
df.shape

(7043, 25)

In [ ]:
# Under 80 (percentage)
total_charges_under80.PhoneService.value_counts() / total_charges_under80.shape[0]

Yes    0.892308
No     0.107692
Name: PhoneService, dtype: float64

In [ ]:
# Above 80 (percentage)
total_charges_above80.PhoneService.value_counts() / total_charges_above80.shape[0]

Yes    0.975936
No     0.024064
Name: PhoneService, dtype: float64

Only 2% of the top 20% (high LTV) did not use the phone service. Lower 80% (LTV) had approximately 11% of individuals who did not use the phone service. That is roughly five times more than the highest 20%. If you do not have a phone connection, you are more likely to be in the bottom 80% of the LTV. We can attempt to sell this service to more people as a company.

##Multiple Lines

In [ ]:
total_charges_under80_use_phone = total_charges_under80.query("PhoneService == 'Yes'")

total_charges_under80_use_phone.MultipleLines.value_counts() / total_charges_under80_use_phone.shape[0]

No     0.590705
Yes    0.409295
Name: MultipleLines, dtype: float64

In [ ]:
total_charges_above80_use_phone = total_charges_above80.query("PhoneService == 'Yes'")

total_charges_above80_use_phone.MultipleLines.value_counts() / total_charges_above80_use_phone.shape[0]

Yes    0.832877
No     0.167123
Name: MultipleLines, dtype: float64

84% of individuals in the top 20% of LTV who used phone service used multiple lines, which is twice the proportion of people in the bottom 80% LTV, which is around 40%. Having numerous lines will bring you closer to the top 20% of the LTV.


##Internet Service

In [ ]:
# Under 80 (percentage)
total_charges_under80.InternetService.value_counts() / total_charges_under80.shape[0]

Fiber optic    0.640803
DSL            0.283612
No             0.075585
Name: InternetService, dtype: float64

In [ ]:
# Above 80 (percentage)
total_charges_above80.InternetService.value_counts() / total_charges_above80.shape[0]

Fiber optic    0.906417
DSL            0.093583
Name: InternetService, dtype: float64

Everyone in the top 20% of LTV had access to the internet. On the other hand, 8% of individuals in the low 80th percentile of LTV did not have access to the internet. In addition, 90% of the top 20% (high LTV) had fibre optic, compared to only 64% of the low LTV. As a result, if we want to increase customer LTV, we should suggest using internet and a Fiber Optic connection.


##Other services

In [ ]:
import numpy as np

In [ ]:
total_charges_under80_use_internet = total_charges_under80.query('InternetService!="No"')
proportion_internet_sub_service_under80 = np.array([total_charges_under80_use_internet.query('OnlineSecurity=="Yes"').shape[0]/total_charges_under80_use_internet.shape[0],
total_charges_under80_use_internet.query('TechSupport=="Yes"').shape[0]/total_charges_under80_use_internet.shape[0],
total_charges_under80_use_internet.query('OnlineBackup=="Yes"').shape[0]/total_charges_under80_use_internet.shape[0],
total_charges_under80_use_internet.query('DeviceProtection=="Yes"').shape[0]/total_charges_under80_use_internet.shape[0],
total_charges_under80_use_internet.query('StreamingTV=="Yes"').shape[0]/total_charges_under80_use_internet.shape[0],
total_charges_under80_use_internet.query('StreamingMovies=="Yes"').shape[0]/total_charges_under80_use_internet.shape[0]])

In [ ]:
proportion_internet_sub_service_under80

array([0.1316932 , 0.13603473, 0.22720695, 0.23516643, 0.37698987,
       0.3813314 ])

In [ ]:
total_charges_above80_use_internet = total_charges_above80.query('InternetService!="No"')
proportion_internet_sub_service_above80 = np.array([total_charges_above80_use_internet.query('OnlineSecurity=="Yes"').shape[0]/total_charges_above80_use_internet.shape[0],
total_charges_above80_use_internet.query('TechSupport=="Yes"').shape[0]/total_charges_above80_use_internet.shape[0],
total_charges_above80_use_internet.query('OnlineBackup=="Yes"').shape[0]/total_charges_above80_use_internet.shape[0],
total_charges_above80_use_internet.query('DeviceProtection=="Yes"').shape[0]/total_charges_above80_use_internet.shape[0],
total_charges_above80_use_internet.query('StreamingTV=="Yes"').shape[0]/total_charges_above80_use_internet.shape[0],
total_charges_above80_use_internet.query('StreamingMovies=="Yes"').shape[0]/total_charges_above80_use_internet.shape[0]])

In [ ]:
proportion_internet_sub_service_above80

array([0.30213904, 0.32620321, 0.55882353, 0.58823529, 0.78342246,
       0.77807487])

Both streaming movie and streaming television rates are around 80% in the top 20% LTV, while they are around 40% in the bottom 80% LTV. Device security and online backup are around 55% for the top 20% LTV, compared to only around 22% for the low LTV. If we want to increase LTV, we would like to offer streaming TV and movies to customers, as well as online backup and device protection.


## 3. For those who churned what is the proportion of each kind of contract?


In [ ]:
# Under 80 (percentage)
total_charges_under80.Contract.value_counts() / total_charges_under80.shape[0]

Month-to-month    0.959197
One year          0.036789
Two year          0.004013
Name: Contract, dtype: float64

In [ ]:
# Above 80 (percentage)
total_charges_above80.Contract.value_counts() / total_charges_above80.shape[0]

Month-to-month    0.590909
One year          0.296791
Two year          0.112299
Name: Contract, dtype: float64

## 4. For those who did not churn what is the proportion of each kind of contract?

In [ ]:
## Extract those who unsubscribed the services (Churn = Yes)

paying_df = pd.read_sql("select * from churn_all where Churn = 'No'", conn)

In [ ]:
paying_df.head()

CustomerID  Gender SeniorCitizen Partner Dependents       State   Latitude  \
0  7590-VHVEG  Female            No     Yes         No  California  33.973616   
1  5575-GNVDE    Male            No      No         No  California  33.949255   
2  7795-CFOCW    Male            No      No         No  California  34.076259   
3  1452-KIOVK    Male            No      No        Yes  California  34.027337   
4  6713-OKOMC  Female            No      No         No  California  34.008293   

    Longitude  ZipCode PhoneService  ... TechSupport StreamingTV  \
0 -118.249020    90001           No  ...          No          No   
1 -118.246978    90002          Yes  ...          No          No   
2 -118.310715    90004           No  ...         Yes          No   
3 -118.285150    90007          Yes  ...          No         Yes   
4 -118.346766    90008           No  ...          No          No   

  StreamingMovies Tenure        Contract              PaymentMethod  \
0              No      1  Month-to-month           Electronic check   
1              No     34        One year               Mailed check   
2              No     45        One year  Bank transfer (automatic)   
3              No     22  Month-to-month    Credit card (automatic)   
4              No     10  Month-to-month               Mailed check   

  PaperlessBilling MonthlyCharges  TotalCharges Churn  
0              Yes          29.85         29.85    No  
1               No          56.95        1889.5    No  
2               No          42.30       1840.75    No  
3              Yes          89.10        1949.4    No  
4               No          29.75         301.9    No  

[5 rows x 25 columns]

In [ ]:
print(paying_df['TotalCharges'].dtype)

object


In [ ]:
print(paying_df['TotalCharges'].unique())

[29.85 1889.5 1840.75 ... 7362.9 346.45 6844.5]


In [ ]:
paying_df.dropna(subset=['TotalCharges'], inplace=True)

In [ ]:
paying_df['TotalCharges'] = pd.to_numeric(paying_df['TotalCharges'], errors='coerce')

In [ ]:
paying_df.TotalCharges.describe()

count    5163.000000
mean     2555.344141
std      2329.456984
min        18.800000
25%       577.825000
50%      1683.600000
75%      4264.125000
max      8672.450000
Name: TotalCharges, dtype: float64

In [ ]:
# Find the 80th percentile of the data in TotalCharges
paying_df.TotalCharges.quantile(0.8)

4895.85

In [ ]:
# Divide the data by the 80th percentile of the data from the TotalCharges variable
paying_total_charges_under80 = paying_df.query("TotalCharges <= 4895.8")
paying_total_charges_above80 = paying_df.query("TotalCharges > 4895.8")

In [ ]:
# Under 80 (percentage)
paying_total_charges_under80.Contract.value_counts() / paying_total_charges_under80.shape[0]

Month-to-month    0.500726
Two year          0.253269
One year          0.246005
Name: Contract, dtype: float64

In [ ]:
# Above 80 (percentage)
paying_total_charges_above80.Contract.value_counts() / paying_total_charges_above80.shape[0]

Two year          0.572120
One year          0.280736
Month-to-month    0.147144
Name: Contract, dtype: float64

- If we want to stop churning, we'd like to offer people contracts for at least a year, preferably two. 
- We have 57% of people in the top 20% LTV who remained in the company with two-year contracts, while only 11% of people in the top 20% had a two-year contract and left the company.
- 95% of those who left the business and are in the low 80% have a month-to-month contract, compared to only 50% of those in the low 80% who remained. One of the reasons for people leaving the business is that they have a month-to-month contract.